mythril/laser/ethereum/state.py

```diff
@@ -135,7 +135,9 @@ class MachineStack(list):
     """
     STACK_LIMIT = 1024
 
-    def __init__(self, default_list=[]):
+    def __init__(self, default_list=None):
+        if default_list is None:
+            default_list = []
```

Comment:
We don’t need that here right? As default_list doesn’t change here and the constructor is of one time usage for the class instance.correct me if I am wrong

Reply:
`default_list=[]` in the function header is the same data structure globally - across all `MachineStack` objects. An example to show the dependencies across objects:
```python
In [3]: class Foo(object):
   ...:     def __init__(self, default_list=[]):
   ...:         self.my_list = default_list
   ...:     def append(self, v):
   ...:         self.my_list.append(v)
   ...:         
In [4]: f = Foo()

In [5]: f.append(42)

In [6]: f.my_list
Out[6]: [42]

In [7]: f2 = Foo()

In [8]: f2.append(1337)

In [9]: f2.my_list
Out[9]: [42, 1337]
```

As you can see, even though we have two separate instances, the state of `f` is dragged into `f2`, because the default arguments of `Foo.__init__` is evaluated only once - when the class is defined. Basically all instances of `Foo` work on the same list pointer. So it's not a problem limited to the constructor, but to everything that touches `default_list` directly or indirectly.

Reply:
Reasons why I hate python seems to keep on growing

Reply:
It's a pretty language once you've figured out its kinks.. :laughing: